File: dsa/dfs/802_Find_Eventual_Safe_States.java
package dfs;
import java.util.*;

class Solution {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        int graphSize = graph.length;
        boolean[] visited = new boolean[graphSize];
        boolean[] path = new boolean[graphSize];
        List<Integer> result = new ArrayList<Integer>();
        for (int i = 0; i < graphSize; i++){
            if (!visited[i]){
                dfs(graph, i, visited, path);
            }
        }
        
        for (int i = 0; i < graphSize; i++){
            if (!path[i]){
                result.add(i);
            }
        }
        return result;
    }

    public boolean dfs(int[][] graph, int curIdx, boolean[] visited, boolean[] path){
        visited[curIdx] = true;
        path[curIdx] = true;
        for (int i = 0; i < graph[curIdx].length; i++){
            if (!visited[graph[curIdx][i]]){
                if (!dfs(graph, graph[curIdx][i], visited, path)){
                    return false;
                }
            }
            
            if (path[graph[curIdx][i]]){
                return false;
            }
        }
        path[curIdx] = false;
        return true;
    }
}

File: dsa/13.dfs/802_Find_Eventual_Safe_States.py
import typing as t

class Solution:
    def eventualSafeNodes(self, graph: t.List[t.List[int]]) -> t.List[int]:
        visited: t.List[bool] = [False] * len(graph)
        path: t.List[bool] = [False] * len(graph)

        for ind in range(0, len(graph), 1):
            if visited[ind]:
                continue 
            self.dfs(graph, ind, visited, path)

        return [ind for ind in range(0, len(path), 1) if not path[ind]]

    def dfs(self,
            graph: t.List[t.List[int]], 
            cur_node: int, 
            visited: t.List[bool], 
            path: t.List[bool]) -> bool:
        
        visited[cur_node] = True
        path[cur_node] = True

        for adj_node in graph[cur_node]:
            if not visited[adj_node]:
                if not self.dfs(graph, adj_node, visited, path):
                    return False

            if path[adj_node]:
                return False
        
        path[cur_node] = False
        return True
